torch/csrc/dynamo/eval_frame.c

```diff
@@ -314,31 +396,228 @@ static void destroy_cache_entry(CacheEntry* e) {
+  // 3) CacheEntryWrapper - This contains the cache_entry for the frame.
+  // 4) A dict from nn module to the CacheEntryWrapper.
+  // Look in get_cache_entry comments for more details.
+
+  // TODO(jansel): would it be faster to bypass this?
+
+  // Case 1 and 2 - Short circuit for SKIP_CODE.
+  if (cache_entry == NULL || cache_entry == SKIP_CODE) {
+    set_cache_entry_on_code(code, cache_entry);
+    return;
+  }
+  // Case 3
+
+  if (nn_module == NULL) {
+    CacheEntryWrapper *cache_entry_wrapper = (CacheEntryWrapper*) PyObject_CallObject((PyObject *) &CacheEntryWrapperType, NULL);
+    cache_entry_wrapper->cache_entry = cache_entry;
+    _PyCode_SetExtra((PyObject*)code, cache_entry_extra_index, (PyObject*) cache_entry_wrapper);
+    return;
+  }
+  DEBUG_CACHE("Setting cache entry for a nn module %p", nn_module);
+
+  // Case 4
+  PyObject* nn_module_to_cache_entry_map = NULL;
+  _PyCode_GetExtra((PyObject*)code, cache_entry_extra_index, (void*)&nn_module_to_cache_entry_map);
+
+  if (nn_module_to_cache_entry_map == NULL) {
+    nn_module_to_cache_entry_map = PyDict_New();
+  }
+
+  PyObject* nn_module_weakref = PyWeakref_NewRef(nn_module, NULL);
+  // TODO - Do I need to DECREF nn_module here? Is it a stolen reference?
+  CacheEntryWrapper *cache_entry_wrapper = (CacheEntryWrapper*) PyObject_CallObject((PyObject *) &CacheEntryWrapperType, NULL);
+  cache_entry_wrapper->cache_entry = cache_entry;
+  PyDict_SetItem(nn_module_to_cache_entry_map, nn_module_weakref, (PyObject*) cache_entry_wrapper);
+  _PyCode_SetExtra((PyObject*)code, cache_entry_extra_index, nn_module_to_cache_entry_map);
+}
+
+inline static PyObject* get_frame_state(PyCodeObject* code, PyObject* nn_module) {
+  // Read get_cache_entry to understand the cache structure. For frame_state, we
+  // have to replicate similar hierarchy. There are 3 possible values.
+  // 1) NULL - First time accessed the extra segment.
+  // 2) Not a NN module method frame - just return PyObject
+  // 3) NN module method frame - access the dict of nn module to frame state.
+
   PyObject* extra = NULL;
   _PyCode_GetExtra((PyObject*)code, dynamic_frame_state_extra_index, (void*)&extra);
-  return extra;
+
+  // Case 1 and 2
+  if (nn_module == NULL || extra == NULL) {
+    return extra;
+  }
+
+  // case 3.
+  PyObject* nn_module_weakref = PyWeakref_NewRef(nn_module, NULL);
+
+  // If the cache is empty, return a null object.
+  PyObject* nn_module_to_cache_entry_map = extra;
+  if (PyDict_GetItem(nn_module_to_cache_entry_map, nn_module_weakref) == NULL) {
+    return NULL;
+  }
+  // Find the frame state.
+  return PyDict_GetItem(nn_module_to_cache_entry_map, nn_module_weakref);
 }
 
-inline static void set_frame_state(PyCodeObject* code, PyObject* extra) {
-  // TODO(jansel): would it be faster to bypass this?
-  _PyCode_SetExtra((PyObject*)code, dynamic_frame_state_extra_index, extra);
+inline static void set_frame_state(PyCodeObject* code, PyObject* extra, PyObject* nn_module) {
```

Comment:
You now have a dedicated `destroy_frame_state`; is it now illegal to pass NULL extra here?